### DISCUSSION TITLE
Add support for Mongodb

### DISCUSSION THREAD

**Discussion by dcousens** [keystonejs/keystone#7511] [Feature requests]
With Prisma supporting Mongodb,  will, and or when will Keystone 6 be adding support?

**Comment by leopoldkristjansson** [comment]
If I'm not mistaken it is supported, since Prisma supports it. There is just no documentation on it specifically.

**Reply by frederik** [reply]
@leopoldkristjansson That is unfortunately not true with the latest release on npm. At this point there is a check that does not include mongodb (on the main branch). If there's another way, I'd love to know.

[  if (!['postgresql', 'sqlite'].includes(config.db.provider)) {](https://github.com/keystonejs/keystone/blob/ad1e5a6972ae84515e1490070979bd6c0710a80b/packages/core/src/lib/config/initConfig.ts#L11)

**Reply by leopoldkristjansson** [reply]
OK, interesting. According to the Keystone Slack channel this is still WIP, but Prisma now allows programmatic creation of a MongoDB database (which apparently was a prerequisite).

https://github.com/keystonejs/keystone/pull/6102#issuecomment-1107455465

https://github.com/prisma/prisma/issues/8582

**Comment by mbrimmer83** [comment]
Any updates on when?

> With Prisma supporting Mongodb, will, and or when will Keystone 6 be adding support?

**Comment by robertherber** [comment]
Would be great with an update on this - is there still something missing on the Prisma side that we're still waiting for? And if not - is this hard to implement? I'm under the impression the Prisma abstraction should make supporting it fairly easy :)

**Comment by ttbarnes** [comment]
Any update on this? Starting on a new project that has to integrate with another system that uses MongoDB.

Would love to be able to use keystone with MongoDB. Otherwise have to do some mappings/migration from SQL to MongoDB.

**Comment by suttonle24** [comment]
I'm just going to pile on here. :slightly_smiling_face: I'm currently working away on a project front-end until Keystone 6 supports MongoDB. At least I hope that ends up being the case, as I really prefer Keystone as a platform.
Please provide any available update regarding MongoDB support - thank you!

**Comment by suttonle24** [comment]
I remembered seeing MongoDB listed on their roadmap page, but it's not there now. :skull: 

Looks like I'm not crazy, as the wayback machine confirmed it back in June:
![image](https://user-images.githubusercontent.com/7423411/185557583-9aa1cabd-f922-451e-8839-0d0949bcb0c5.png)

**Comment by dcousens** [comment]
### Update, 19 August 2022
When we decided to make Prisma a core part of Keystone 6, we knew they had a plan to support MongoDB at some stage in the future.

With Prisma recently shipping General Availability support for MongoDB, we have taken the time to consider the feedback from the Keystone community, and dive into the details of Prisma’s implementation in order to come up with a balanced evaluation of the pros and cons for adding support in Keystone.

Unfortunately, the outcome of our investigation is that **we have decided not to support MongoDB at this point in time.**
Here are a few of the reasons for how we came to this decision:

- **Inconsistent migration support** – [Prisma has no plans for supporting migrations with Mongo](https://github.com/prisma/prisma/issues/7305). Without migrations, a key part of the Keystone DX would be impacted.  For now we'd rather have one easy-to-understand story when it comes to migrations in typical workflows.
- **Mandatory transactions** – Prisma’s integration with MongoDB requires support for transactions. This is not the default setup for Mongo, and [additional database subtleties in configuration, like replica sets](https://github.com/prisma/docs/pull/3017), would be required to make it work in Keystone. This added complexity works against the simple “out-of-the-box” DX we are striving for in Keystone 6.
- **Roadmap velocity**  – While Prisma is doing most of the heavy abstraction work when it come to databases in Keystone 6, none-the-less, maintaining support for relational _and_ document databases in the Keystone project will result in an additional maintenance strain.

As a result, we feel that by maintaining an exclusive focus on what we can support, in the short to medium term, that not adding support for MongoDB will leave us positioned to deliver other key features at a faster rate, to a higher standard, and without negatively impacting the Keystone-6 DX that we are working really hard to achieve.
With the recent [inclusion of MySQL](https://github.com/keystonejs/keystone/pull/7538), end developers now have first-class support for the top 3 relational databases as voted by the community in last year's developer survey.

Accordingly, and as [recently updated in our roadmap](https://github.com/keystonejs/keystone/pull/7714#issue-1304310381)  (and pointed out by @suttonle24) we have removed MongoDB from our [public roadmap](https://keystonejs.com/updates/roadmap).

This environment is always moving forward,  and if the opportunity arises where we can integrate MongoDB without compromising on what we are working towards - we will happily re-evaluate this position.

We know this isn't the news that many of you wanted, but hopefully it's only a temporary position as we move onwards and upwards!

Team Keystone

**Reply by suttonle24** [reply]
Really appreciate the detailed response @dcousens . I do completely understand the need for limiting scope to help drive up overal quality of what you _**do**_ deliver. :) Thank you for the answer there.

Looks like it's on to DB migration then!

**Comment by jbool24** [comment]
I hope this position changes in the future as one of the main use cases for Mongo Support would be to build internal dashboards to help non-technical teams self manage data with easy to use Input Forms on a front end. Ideally, some input Forms that could be programed with flow control and logic to maintain existing business rules like a side-car BaaS for internal non-tech teams. Something similar in the RDB world is Directus or Strapi.

There is a huge gap in the NoSQL world that Keystone could totally fulfill (and create consulting business around to support the project. 😉  I work with start-ups on 99% of my projects and this could be a game changer for ongoing maintenance. Startups often choose MongoDB when the domain is not yet fully "realized" or fleshed out because of its schema-less flexibility. No migrations (or at least the concern is pushed up a layer into code). When a startup is meant to launch and start production support but doesn't yet have the resources to build an Tech Support internal UI to admin "stuff" with the data, Keystone could be the perfect tool to quickly get a Dash and set of Forms to update what's needed without the need to run scripts or directly manipulate data in a DMS like Studio 3T or Compass.

I'd love to contribute further to the conversation to hopefully put this feature back on the roadmap.

PS. If anyone knows of other software thats good for this use case and compatible with MongoDB, I'd love to know.

**Reply by Mshah-98** [reply]
Hey have you found anything that would work with mongoDB? I am one of the startups you're talking about haha

**Reply by jbool24** [reply]
I've long moved on from Keystone. Checkout [Parse Server](https://parseplatform.org/), not MVC but one of the only ones I found that supports MongoDB. You can build a headless client on top of that with any flavor of the week front end (React, Vue, Astro, etc etc).

As another option, you could also just build on top of Mongo Atlas. They have a VERY generous free tier and if you just want to get a product up an running to validate proof-of-value and market acceptance just do that.

Just a word of caution, once you progress to a growth/scale-up stage, you probably want to migrate off of their serverless stack to something you control. You probably won't need to do that for some time though.

I've done this for several clients now and it has proved to be a very good path; initial launch on Atlas and then scale with something custom only after we prove viability.

**Comment by hungtrungthinh** [comment]
I regret that Keystone kicked Mongo out of the roadmap. Most of the projects we start with require MongoDB and essential components with nested models, where we can design apps with a more significant amount of document data processing. I regret that Keystone now cannot be combined with Nextjs 13 when we create a great product with excellent performance because Keystone has excluded Mongo from the roadmap, making me feel pain.

**Comment by blueonrails** [comment]
It is a great pity that keystone6 does not support MongoDB. In a way it is breaking support for projects using keystone5 with Mongo, as they cannot upgrade to keystone6. 

If at some point the team plans to support it, I could help in the development, as I am a bit affected by the lack of mongo support.

Thank you very much for this work on the CMS, it's great 👏

**Comment by reslear** [comment]
any news?
